src/web/lib/useVisibility.ts

```diff
@@ -0,0 +1,30 @@
+import { useEffect, useState, useRef } from 'react';
+
+const useVisibility = (threshold: number) => {
+    const [isVisible, updateVisibility] = useState<boolean>(false);
+    const [node, setNode] = useState<HTMLElement | null>(null);
+
+    const observer = useRef<IntersectionObserver | null>(null);
+
+    useEffect(() => {
+        if (observer.current) observer.current.disconnect();
+
+        observer.current = new window.IntersectionObserver(
+            ([entry]) => updateVisibility(entry.isIntersecting),
+            {
+                root: null,
+                threshold,
+            },
+        );
+
+        const { current: currentObserver } = observer;
+
+        if (node) currentObserver.observe(node);
+
+        return () => currentObserver.disconnect();
+    }, [node, threshold]);
+
+    return [setNode, isVisible];
```

Comment:
I think, by convention, these should be returned the other way around. It' doesn't matter but might trip someone up later on!